appvia/kore
# Bind services to clusters

## What

We had a discussion about the future of services and we think we've come up with a solution which will simplify things in the future and give us a more constrained scope:
 * all services will be owned by a single cluster (which means the cluster field will be required)
 * there will be an option to share a service with other clusters (which will only allow to manage service access (service credentials in an other cluster) - not part of this story
 * there will be two types of services like now: cloud services and application services (in-cluster services). One technical distinction is that application services will need a cluster namespace, while cloud services won't.

Technically an application service is where `kore.appvia.io/platform=Kubernetes`, everything else is a cloud service.

### Changes

 * New cluster page layout with tabs
 * The cloud services were moved to the cluster page
 * Tab permalinks on the cluster page (thanks to @davet1985)
 * The service page was moved under the cluster URL (thanks to @davet1985)
 * Add a ServiceAccessEnabled flag to service kinds (if false, you can't create service accesss)
 * Add a ServiceAccessDisabled flag to service plans (if true, you can't create service access)
 * Strict handling of the feature gates everywhere
    * the services menu on the left should only show up when the services feature gate is enabled
    * the services config page should not show app service kinds if the feature gate is not enabled.
 * The Dummy service will create a service kind for AWS, GCP, Azure and Kubernetes for easier testing
 * Add cluster and namespace options to `kore create service`
 * various smaller fixes/improvements (see commits)

### New cluster page layout

![Screenshot 2020-06-09 at 18 17 05](https://user-images.githubusercontent.com/868104/84189204-ce35d800-aa8c-11ea-9966-fbd4a5efee06.png)

### Testing

1. Make sure the `application_services` feature gate is enabled on the UI. If you use `make run` or `make run-prod`, it is by default.
1. Install the dummy service provider
    ```
    kore apply -f examples/service-provider-dummy.yml
    ```
1. Go through the UI and test all service-related functionality.
1. Disable `application_services` on the UI and check again.
1. Disable `services` on the UI and check again.

Functionality to check:
 * http://localhost:3000/configure/services - enable/disable services, manage services
 * create/update/delete a cloud service
 * create/update/delete an application service
 * create/update/delete service access
 * create a cloud service using the Kore CLI
    ```
    $ kore create service cloud-test --plan dummy-aws-default --cluster kore -t kore-admin
    ```
 * create an app service using the kore CLI
    ```
    kore create service app-test --plan dummy-kubernetes-default --cluster kore -t kore-admin -n app-test-ns
    ```
